### PR TITLE
Aggregate list of all uses for a single species

### PR DESCRIPTION
**Pull request by sandreae**
Introduces a widget which aggregates and displays all uses for a particular species.

image

The query logic required for this PR required some thought, and in the end I couldn't find a perfect solution. It's a little ugly (see below) but I think it is the best we can do, and there are positives to the final approach as well. I'm not against exploring different ideas, but I can't think of any myself :sweat_smile: 

## `UsedFor` document collection queries

As there is no direct relation between a species and it's uses we need to make 2 collection queries in order to 1) get all sightings associated with the species we are viewing 2) paginate over all uses which have been attached to any of the sightings (using an `in` filter). This would be fine, except the query required for step 2) will return a collection containing (potentially many) duplicate uses. This is because a unique "UsedFor" document is created whenever we attach a use to a sighting, even if it is actually "the same" use (eg. assigning "Put it on a cut" as a use for sighting A creates a new UsedFor document, doing the same for sighting B creates _another_ UsedFor document). This means we need to de-duplicate items in the returned collections so that we don't display a list containing many duplicates to the user.

_This is the same issue discovered and discussed when we introduced the `UsedForField` I'm just explaining it here again so we're clear on the problem at hand._

Although the idea of introducing a join document to our schema designs was floated as a way to remove this duplication of uses documents, actually it only solves the problem in one case (the UsedForField query) but not this one. So I don't really see the point of making a change to our schema if it doesn't solve the problem everywhere. Unfortunately I don't see any solution which solves it in both cases. We would need to be able to query reverse relations in our GraphQL api I think.

## But on the bright side

Those are the negatives, but there are positives too! Mainly, that with minor changes, I was able to re-use existing queries `allSightingsQuery` and `allUsesQuery` to compose the exact query we need here. Also re-use of the widget for paginating over (and de-duplicating) the uses was possible. So this PR is very small now. 

## Implementation
- introduced `FetchAll` wrapper which fetches all sightings (collects all pages) of the desired species and passes them on to a child widget via a custom builder method
- inside this pop in the existing `DeduplicatedUsedForTagsList` which filters by the received collection of sighting ids and paginates over the results (user clicks more for more).

### CODE REVIEW

File: packages/app/lib/ui/widgets/pagination_list.dart [hunk 1]
```diff
@@ -21,12 +21,22 @@ typedef ContainerBuilder<T> = Widget Function(
...
+typedef FetchMoreOverride<T> = bool Function(PaginatedCollection<T>);
+
 class PaginationBase<T> extends StatelessWidget {
   final ContainerBuilder<T> builder;
   final Paginator<T> paginator;
 
+  /// Optional function which can override the default behaviour regarding when
+  /// the next page of documents are requested. By default this only happens on
+  /// user interaction. With this method you could, for example, continue to call fetchNext.
+  final FetchMoreOverride<T>? fetchMoreOverride;
```
Comment (sandreae) [inline packages/app/lib/ui/widgets/pagination_list.dart:33]: Optional method which can be used to override default "fetch more" behaviour.

File: packages/app/lib/ui/widgets/pagination_list.dart [hunk 2]
```diff
@@ -201,6 +215,30 @@ class SliverPaginationBase<T> extends StatelessWidget {
   }
 }
 
+class FetchAll<T> extends StatelessWidget {
```
Comment (sandreae) [inline packages/app/lib/ui/widgets/pagination_list.dart:218]: New widget which fetches all pages of a collection query in one go, and passes the whole collection onto it's child via a custom builder method.

File: packages/app/lib/ui/widgets/species_uses_aggregate.dart [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      constraints: const BoxConstraints(maxHeight: 300),
+      child: SingleChildScrollView(
+        child: Container(
+            padding:
+                const EdgeInsets.symmetric(vertical: 10.0, horizontal: 18.0),
+            child: FetchAll(
```
Comment (sandreae) [inline packages/app/lib/ui/widgets/species_uses_aggregate.dart:53]: We use the `FetchAll` widget here to get all sightings for a particular species and then pass the result onto `UsedForTagsList` which.....

File: packages/app/lib/ui/widgets/used_for_dedup_tag_list.dart [hunk 4]
```diff
@@ -3,132 +3,43 @@
...
 
 class DeduplicatedUsedForTagsList extends StatelessWidget {
   final PaginationBuilder<UsedFor> builder;
   final Paginator<UsedFor> paginator;
+  final minimumPageSize = 10;
 
   const DeduplicatedUsedForTagsList(
       {super.key, required this.builder, required this.paginator});
 
+  bool _fetchMoreOverride(PaginatedCollection<UsedFor> data) {
```
Comment (sandreae) [inline packages/app/lib/ui/widgets/used_for_dedup_tag_list.dart:17]: We override default behaviour here as we want to remove unavoidable duplicate "uses" from the collection query results.

File: packages/app/lib/ui/screens/species.dart [hunk 5]
```diff
@@ -115,6 +116,8 @@ class _SpeciesProfileState extends State<SpeciesProfile> {
           ])),
           RelatedSightings(id: species.id),
           const SliverToBoxAdapter(child: SizedBox(height: 20.0)),
+          SliverToBoxAdapter(child: SpeciesUsesAggregate(id: species.id)),
+          const SliverToBoxAdapter(child: SizedBox(height: 20.0)),
```
Comment (adzialocha) [inline packages/app/lib/ui/screens/species.dart:120]: Looking at the wireframes I believe that the "sightings grid" comes last, then you can move your widget into the `SliverList` which accepts only widgets aka you don't need an adapter then 😅

Comment (sandreae) [reply packages/app/lib/ui/screens/species.dart]: Nice! No adaptor :fist_raised:

**Pull request merged**